File: services/shared/contracts.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from services.shared.enums import ChatMode, FileStatus, PendingRequestStatus, UploadAndAskScope


class ContractModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class CitationResponse(ContractModel):
    file_id: str = Field(alias="fileId")
    file_name: str = Field(alias="fileName")
    page: int | None = None
    sheet_name: str | None = Field(default=None, alias="sheetName")
    section_heading: str | None = Field(default=None, alias="sectionHeading")
    snippet: str
    chunk_id: str = Field(alias="chunkId")
    score: float


class FileSummary(ContractModel):
    id: str
    workspace_id: str = Field(alias="workspaceId")
    file_name: str = Field(alias="fileName")
    status: FileStatus
    uploaded_at: datetime = Field(alias="uploadedAt")


class FileDetail(FileSummary):
    mime_type: str = Field(alias="mimeType")
    size_bytes: int = Field(alias="sizeBytes", ge=0)
    object_key: str = Field(alias="objectKey")
    parser_used: str | None = Field(default=None, alias="parserUsed")
    error_message: str | None = Field(default=None, alias="errorMessage")


class UploadResponse(ContractModel):
    file_id: str = Field(alias="fileId")
    status: FileStatus
    message: str


class ChatQueryRequest(ContractModel):
    workspace_id: str = Field(alias="workspaceId")
    chat_session_id: str | None = Field(default=None, alias="chatSessionId")
    mode: ChatMode
    query: str
    scope: UploadAndAskScope
    file_ids: list[str] = Field(default_factory=list, alias="fileIds")


class ChatQueryResponse(ContractModel):
    request_id: str = Field(alias="requestId")
    status: PendingRequestStatus
    answer: str | None = None
    citations: list[CitationResponse] = Field(default_factory=list)
    pending_request_id: str | None = Field(default=None, alias="pendingRequestId")


class PendingUploadAndAskRequestState(ContractModel):
    request_id: str = Field(alias="requestId")
    workspace_id: str = Field(alias="workspaceId")
    status: PendingRequestStatus
    scope: UploadAndAskScope
    query: str
    file_ids: list[str] = Field(default_factory=list, alias="fileIds")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

File: packages/shared-types/src/contracts.ts
import type { ChatMode, FileStatus, PendingRequestStatus, UploadAndAskScope } from "./enums";

export interface CitationResponse {
  fileId: string;
  fileName: string;
  page: number | null;
  sheetName: string | null;
  sectionHeading: string | null;
  snippet: string;
  chunkId: string;
  score: number;
}

export interface FileSummary {
  id: string;
  workspaceId: string;
  fileName: string;
  status: FileStatus;
  uploadedAt: string;
}

export interface FileDetail extends FileSummary {
  mimeType: string;
  sizeBytes: number;
  objectKey: string;
  parserUsed: string | null;
  errorMessage: string | null;
}

export interface UploadResponse {
  fileId: string;
  status: FileStatus;
  message: string;
}

export interface ChatQueryRequest {
  workspaceId: string;
  chatSessionId: string | null;
  mode: ChatMode;
  query: string;
  scope: UploadAndAskScope;
  fileIds: string[];
}

export interface ChatQueryResponse {
  requestId: string;
  status: PendingRequestStatus;
  answer: string | null;
  citations: CitationResponse[];
  pendingRequestId: string | null;
}

export interface PendingUploadAndAskRequestState {
  requestId: string;
  workspaceId: string;
  status: PendingRequestStatus;
  scope: UploadAndAskScope;
  query: string;
  fileIds: string[];
  createdAt: string;
  updatedAt: string;
}
